File: 0450-delete-node-in-a-bst/0450-delete-node-in-a-bst.java
class Solution {
    public TreeNode deleteNode(TreeNode root, int key) {
        
        if(root == null)
            return null;
        
        if(root.val > key){
            root.left = deleteNode(root.left, key);
        }
        else if(root.val < key){
            root.right = deleteNode(root.right, key);
        }
        else{
            
            if(root.left == null)
                return root.right;
            else if(root.right == null)
                return root.left;
            else{
                
                int newData = findInorderSuccessor(root.right);
                root.val = newData;
                
                root.right = deleteNode(root.right, newData);
            }
            
            
        }
        
        return root;
    }
    
    public int findInorderSuccessor(TreeNode root){
        
        int data = root.val;
        
        while(root.left != null){
            data = root.left.val;
            root = root.left;
        }
        return data;
    }
}

File: 0450-delete-node-in-a-bst/0450-delete-node-in-a-bst.cpp
class Solution {
public:
    TreeNode* deleteNode(TreeNode* root, int key) {
        
        if(root == NULL)
            return NULL;
        
        if(root->val > key){
            root->left = deleteNode(root->left, key);
        }
        else if(root->val < key){
            root->right = deleteNode(root->right, key);
        }
        else{
            if(root->left == NULL)
                return root->right;
            else if(root->right == NULL)
                return root->left;
            
            int newData = findInorderSuccessor(root->right);
            root->val = newData;
            
            root->right = deleteNode(root->right, newData);
        }
        
        return root;
    }
    
    int findInorderSuccessor(TreeNode* root){
        
        int data = root->val;
        while(root->left != NULL){
            data = root->left->val;
            root = root->left;
        }
        
        return data;
    }
};